### PR TITLE
[android] remove ColorStateList(int[][],int[]) usage

### PR DESCRIPTION
**Pull request by jonathanpeppers**
`dotnet trace` output of `dotnet new maui` on a Pixel 5 device shows:

    16.63ms microsoft.maui!Microsoft.Maui.Platform.ColorStateListExtensions.CreateDefault(int)
    16.63ms mono.android!Android.Content.Res.ColorStateList..ctor(int[][],int[])
    11.38ms mono.android!Android.Runtime.JNIEnv.NewArray

Reviewing, the Android API for `ColorStateList`, the C# binding is:

    public unsafe ColorStateList (int[][]? states, int[]? colors)
        : base (IntPtr.Zero, JniHandleOwnership.DoNotTransfer)
    {
        //...
        IntPtr native_states = JNIEnv.NewArray (states);

There is a general performance problem in using C# arrays to call into
Java. We have to create a Java array and copy each array element over
to the Java array. We slightly improved upon this scenario in
xamarin/xamarin-android#6870, but the array duplication remains a
problem.

To solve this:

1. Make the `ColorStateList` ctor a "banned API" in C#.

2. Delete `ColorStates.cs`. Create a private `ColorStates` class in Java.

3. Create Java methods for creating `ColorStateList` objects.

I also removed `ColorExtensions.IsOneColor()`, in favor of the pattern:

    if (PlatformInterop.CreateEditTextColorStateList(platformPicker.TextColors, textColor.ToPlatform()) is ColorStateList c)
        platformPicker.SetTextColor(c);

This reduces the number of JNI calls in this case.

After these changes, we instead get:

    2.44ms microsoft.maui!Microsoft.Maui.Platform.ColorStateListExtensions.CreateDefault(int)

Which appears to save ~14ms of startup time on a Pixel 5 in `dotnet new maui`.
Other apps that use `Entry`, `CheckBox`, `Switch`, etc. will also see an improvement.

The only place I didn't fix this is in `TabPage`, as it was not a
straightforward port. Since most apps will be using Shell going
forward anyways, we can revisit this in the future, if needed.

### CODE REVIEW

File: src/Compatibility/Core/src/Android/AppCompat/TabbedPageRenderer.cs [hunk 1]
```diff
@@ -1011,7 +1011,10 @@ ColorStateList GetColorStateList(int defaultColor, int checkedColor)
 			states[1] = GetEmptyStateSet();
 			colors[1] = defaultColor;
 
+#pragma warning disable RS0030
+			//TODO: port this usage to Java, if this becomes a performance concern
 			return new ColorStateList(states, colors);
+#pragma warning restore RS0030
```
Comment (jonathanpeppers) [inline src/Compatibility/Core/src/Android/AppCompat/TabbedPageRenderer.cs:1017]: I did not port `TabPage` to use the new methods. Shell should be more common, anyway?

We can fix this in the future if needed.

File: src/Core/src/Platform/Android/ColorStateListExtensions.cs [hunk 2]
```diff
@@ -5,49 +5,25 @@ namespace Microsoft.Maui.Platform
...
 
 		public static ColorStateList CreateEditText(int enabled, int disabled) =>
-			new ColorStateList(ColorStates.EditText, new[] { enabled, disabled });
+			PlatformInterop.GetEditTextColorStateList(enabled, disabled);
 
 		public static ColorStateList CreateCheckBox(int all) =>
 			CreateCheckBox(all, all, all, all);
 
 		public static ColorStateList CreateCheckBox(int enabledChecked, int enabledUnchecked, int disabledChecked, int disabledUnchecked) =>
-			new ColorStateList(ColorStates.EditText, new[] { enabledChecked, enabledUnchecked, disabledChecked, disabledUnchecked });
```
Comment (jonathanpeppers) [inline src/Core/src/Platform/Android/ColorStateListExtensions.cs:39]: Was this a typo before? It was using `ColorStates.EditText` for `CheckBox`.

File: src/Core/src/Platform/Android/ColorStateListExtensions.cs [hunk 3]
```diff
@@ -5,49 +5,25 @@ namespace Microsoft.Maui.Platform
...
-
-		public static ColorStateList CreateSwitch(int all) =>
-			CreateSwitch(all, all, all);
+			PlatformInterop.GetCheckBoxColorStateList(enabledChecked, enabledUnchecked, disabledChecked, disabledUnchecked);
 
 		public static ColorStateList CreateSwitch(int disabled, int on, int normal) =>
-			new ColorStateList(ColorStates.EditText, new[] { disabled, on, normal });
```
Comment (jonathanpeppers) [inline src/Core/src/Platform/Android/ColorStateListExtensions.cs:45]: Was this a typo before? It was using `ColorStates.EditText` for `Switch`.

File: src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellBottomNavViewAppearanceTracker.cs [hunk 4]
```diff
@@ -142,15 +142,7 @@ ColorStateList MakeColorStateList(Color titleColor, Color disabledColor, Color u
 
 		static ColorStateList MakeColorStateList(int titleColorInt, int disabledColorInt, int defaultColor)
 		{
-			var states = new int[][] {
-				new int[] { -R.Attribute.StateEnabled },
-				new int[] {R.Attribute.StateChecked },
-				new int[] { }
-			};
```
Comment (jonathanpeppers) [inline src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellBottomNavViewAppearanceTracker.cs:149]: This `int[][]` was the same as `ColorStates.Switch`, so I just used `ColorStateListExtensions.CreateSwitch()`.

**Comment by jonathanpeppers** [comment]
To test this, I made a layout with:
```xml
<Button Text="enabled" />
<Button IsEnabled="false" Text="disabled" />

<Entry Text="enabled" />
<Entry IsEnabled="false" Text="disabled" />

<Label Text="enabled" />
<Switch />
<Label Text="toggled" />
<Switch IsToggled="true" />
<Label Text="disabled" />
<Switch IsEnabled="false" />
<Label Text="both" />
<Switch IsToggled="true" IsEnabled="false" />

<Label Text="enabled" />
<CheckBox />
<Label Text="checked" />
<CheckBox IsChecked="true" />
<Label Text="disabled" />
<CheckBox IsEnabled="false" />
<Label Text="both" />
<CheckBox IsChecked="true" IsEnabled="false" />

<DatePicker Date="7/1/2022" />
<DatePicker IsEnabled="false" Date="7/1/2022" />
```

I think they look the same. Before:



After:

**Review by mattleibow — APPROVED** [review]
Looks good. Waiting on another set of @PureWeen eyes.

**Review by PureWeen — APPROVED** [review]

**Pull request merged**